### DISCUSSION TITLE
Help me comply with licenses

### DISCUSSION THREAD

**Discussion by ZeroEkkusu** [smartcontractkit/full-blockchain-solidity-course-py#674] [Q&A] [answered]
Gm, frens,

Are there some common ways of complying with the standardized open source licenses among devs, like, where to put license files, giving credits, etc?

I did find [Choose a License](https://choosealicense.com/) and [The Legal Side of Open Source](https://opensource.guide/legal/) helpful, but still have questions.

Here's my [aave_brownie_py](https://github.com/ZeroEkkusu/aave_brownie_py) repo.
I chose the MIT license and added my copyright.

**1. Let's start from the dependencies**

[aave/protocol-v2](https://github.com/aave/protocol-v2) is AGPL-3.0

These files are not in my repo, but they do get fetched (outside the project dir) when compiling. The `LICENSE` file is fetched, as well.
**Q1.1** - Do I need to add anything to my repo, then?

However, I do have `ILendingPool.sol` and `IProtocolDataProvider.sol` inside my `interfaces` directory, both of which are licensed under AGLP-3.0.
**Q1.2** - Should I put a AGPL-3.0 license file in there?
**Q1.3** - There's no copyright notice in these files, but there is one in Aave's `LICENSE`. That's it, right, where should I put it?

**2. Other interfaces**

`AggregatorV3Interface.sol` is MIT - the same as my repo. I already have the license in my root dir.
**Q2** - What should I do here? (The copyright notice situation is the same as above with Aave)

**3. [PatrickAlphaC/aave_brownie_py](https://github.com/PatrickAlphaC/aave_brownie_py)**

Btw, Patrick did not provide any licenses copies / copyright notices besides the repo's `LICENSE`. 
Patrick's repo is MIT licensed. My repo is based on this repo and is heavily modified.
**Q3** - How should I credit Patrick?

**4. I have a few more questions but that's it for now**

**Comment by cromewar** [comment]
@PatrickAlphaC  could you give us a hand o this one?

**Reply by PatrickAlphaC** [reply]
Wow, totally missed this. I'm so sorry!

**Comment by ZeroEkkusu** [comment]
@cromewar delete the discussion, please
The matter is more complicated then I thought

**Reply by cromewar** [reply]
I was doing research on this actually, but didn't fin anything reliable just that each contract file should be treated as independent with it's own license, so you could multiple licenses on your project and each file is ruled by that specific one.

**Comment by ZeroEkkusu** [comment, marked as answer]
Okay, here's [the commit](https://github.com/ZeroEkkusu/aave_brownie_py/commit/c6511cbcd16bc61edf383de60508b98b8e24e3ad) I've made. TL;DR is down there👇

Of course, I did some research, too, and, as @cromewar said, the information isn't very concrete. On top of that, this is mostly legal stuff. However, two things are clear:
- Some licenses are incompatible with one another
> If license A says extended programs must be under license A, and license B says extended programs must be under license B, they have an irreconcilable disagreement.
- Copyleft licenses, such as AGPLv3, are stricter than permissive licenses, such as MIT
![](https://dwheeler.com/essays/floss-license-slide-image.png)

So, here's what I've done:
- Wherever I copied or modified Patrick's code, I added, respectively:
```
Copied from PatrickAlphaC/aave_brownie_py (https://github.com/PatrickAlphaC/aave_brownie_py). See NOTICE.md.
```
```
Modified from PatrickAlphaC/aave_brownie_py (https://github.com/PatrickAlphaC/aave_brownie_py). See NOTICE.md.
```
You can see examples [here](https://github.com/ZeroEkkusu/aave_brownie_py/blob/main/scripts/lending_pool.py).
- Then, I made [`NOTICE.md`](https://github.com/ZeroEkkusu/aave_brownie_py/blob/main/NOTICE.md), with a copy of Patrick's aave_brownie_py license inside, where I stated that the project is based on his (and mentioned it in the `README.md`, as well)
- I isolated the Chainlink and Aave interfaces inside dedicated directories together with their `LICENSE`s
- Finally, I relicensed my repo to AGPLv3 because (I think - see the graphic above) using the MIT license is not an option here, i.e. MIT --> AGLPv3 ✔️ MIT <-- AGLPv3 ❌ because my Aave dependency is AGLPv3, and it's viral

**Reply by cromewar** [reply]
You are awesome, I would like everyone cares this much about the licenses as you do, have you considered writing an article about this?

**Reply by ZeroEkkusu** [reply]
Thank you for the suggestion. Maybe not on this matter, though, since I'm not a lawyer and don't understand it well enough yet. But a repository where others, more knowledgeable than me, can contribute too would be a good option.

Edit: Making that repository Web3 specific is another good idea, as to minimize the scope and maximize the utility of information.

**Reply by n4n0b1t3** [reply]
Wow, that is very thorough! I was contemplating this topic recently too. In many cases, it is not even necessary to have a written copyright notice in documents. GitHub is a good way to prove, that at one point someone came up with something. However, I believe that time to market is the most important thing and getting a user base. When anyone has the time and means to copy your idea, there is little you can do against it, unless you have the money to hire lawyers and as we have seen with Zuck, even this doesn't always work.

**Reply by ZeroEkkusu** [reply]
Those are good points @n4n0b1t3. I'd add marketing to those. In the world when most can code and anyone can fork, you need to set your self apart.

As for the licenses (setting the business aspects aside), you have to comply with terms and conditions.

My approach with `Copied from...` would not cut it. The MIT license requires those big headers whenever you use someone's code, which is inconvinent.

Some of those licences are meant to prevent use in closed source or require special permissions (think $$$).

One good approach is to license the repo under the MIT/Apache dual license - to let the user choose which one to use. The MIT requires headers and is compatible with AGLPv3, while the Apache does not require the headers but also is not compatible with AGPLv3.

I used it for this library: [**🧪 Optimizooor's Lab**](https://github.com/ZeroEkkusu/optimizooors-lab).

**Reply by PatrickAlphaC** [reply]
This is awesome!! So sorry I totally missed this discussion... Looks like yall found a solution here. :) 

Licenses are stuff I've shyed away from and just made everything MIT so people can use whatever. 

Thanks for looking to credit the work!

**Reply by n4n0b1t3** [reply]
@ZeroEkkusu I am impressed, your projects have an aesthetic quality that is hard to find. Would it be OK if we [connect on LinkedIn](https://github.com/n4n0b1t3)?

**Reply by ZeroEkkusu** [reply]
whats linkedin?

<br>

(I'm not on LinkedIn)
